modApiServer/src/org/aion/api/server/pb/ApiAion0.java

```diff
@@ -307,1514 +364,2092 @@ public ApiAion0(IAionChain ac) {
-                        req.getDuration());
-            } catch (InvalidProtocolBufferException e) {
-                LOG.error("ApiAion0.process.unlockAccount exception: [{}]", e.getMessage());
-                return ApiUtil.toReturnHeader(getApiVersion(), Retcode.r_fail_function_exception_VALUE);
-            }
+                        Address addr = Address.wrap(req.getAddress().toByteArray());
 
-            byte[] retHeader = ApiUtil.toReturnHeader(getApiVersion(), Retcode.r_success_VALUE);
-            return ApiUtil.combineRetMsg(retHeader, (byte) (result ? 0x01 : 0x00));
-        }
+                        nonce = this.getNonce(addr);
+                    } catch (InvalidProtocolBufferException e) {
+                        LOG.error("ApiAionA0.process.getNonce exception: [{}]", e.getMessage());
+                        return ApiUtil.toReturnHeader(
+                                getApiVersion(), Message.Retcode.r_fail_function_exception_VALUE);
+                    }
 
-        // Transaction Module
-        case Message.Funcs.f_getBalance_VALUE: {
+                    Message.rsp_getNonce rsp =
+                            Message.rsp_getNonce
+                                    .newBuilder()
+                                    .setNonce(ByteString.copyFrom(nonce.toByteArray()))
+                                    .build();
 
-            if (service != Message.Servs.s_chain_VALUE) {
-                return ApiUtil.toReturnHeader(getApiVersion(), Retcode.r_fail_service_call_VALUE);
-            }
+                    byte[] retHeader =
+                            ApiUtil.toReturnHeader(
+                                    getApiVersion(), Message.Retcode.r_success_VALUE);
+                    return ApiUtil.combineRetMsg(retHeader, rsp.toByteArray());
+                }
+            case Message.Funcs.f_getNrgPrice_VALUE:
+                {
+                    if (service != Servs.s_tx_VALUE) {
+                        return ApiUtil.toReturnHeader(
+                                getApiVersion(), Retcode.r_fail_service_call_VALUE);
+                    }
 
-            byte[] data = parseMsgReq(request, msgHash);
-            BigInteger balance;
-            try {
-                Message.req_getBalance req = Message.req_getBalance.parseFrom(data);
+                    // instantiate nrg price oracle
+                    IAionBlockchain bc = (IAionBlockchain) AionImpl.inst().getBlockchain();
+                    long nrgPriceDefault = CfgAion.inst().getApi().getNrg().getNrgPriceDefault();
+                    long nrgPriceMax = CfgAion.inst().getApi().getNrg().getNrgPriceMax();
+                    NrgOracle.Strategy oracleStrategy = NrgOracle.Strategy.SIMPLE;
+
+                    this.nrgOracle = new NrgOracle(bc, nrgPriceDefault, nrgPriceMax, oracleStrategy);
+                    long nrg = this.getRecommendedNrgPrice();
```

Comment:
Web3Aion has the same logic, please considering move the NrgOracle logic in both web3 and java api to the abstract class ApiAion